### DISCUSSION TITLE
Markdown-It Plugins

### DISCUSSION THREAD

**Discussion by justneedham** [markdoc/markdoc#313] [Q&A] [answered]
Since MarkDoc leverages `markdown-it` under the hood, it is possible to use markdown-it plugins? I've been over the documentation a couple times and I didn't see anything that suggested one way or the other. 

What I'm after is the ability to get closer to Github flavored markdown. (Emoji support/mermaid diagrams). Although I suspect that latter could be added manually via a custom tag?

**Comment by rpaul-stripe** [comment, marked as answer]
We don't officially support using `markdown-it` plugins inside of Markdoc. We largely treat `markdown-it` as an implementation detail and may switch to a different tokenizer at some point in the future. It's technically possible to modify or replace the Markdoc tokenizer to use a custom `markdown-it` configuration with plugins, but the part of Markdoc that converts the `markdown-it` tokens to Markdoc AST nodes doesn't recognize non-standard tokens.

We generally recommend using Markdoc tags and transforms for this kind of extensibility, but if you absolutely must use a `markdown-it` plugin, you can do it by extending the tokenizer and then converting all of the custom tokens to standard ones that the Markdoc parser recognizes. Here's a working example for the emoji plugin:

```
import Markdoc from '@markdoc/markdoc';
import emoji from 'markdown-it-emoji';

const example = `This is a test :smile:`;

class CustomTokenizer extends Markdoc.Tokenizer {
  constructor(config) {
    super(config)
    this.parser.use(emoji);
  }
}

function convertToken(token) {
  if (token.type === 'emoji') token.type = 'text';
  token.children?.forEach(convertToken);
}

const tokenizer = new CustomTokenizer();
const tokens = tokenizer.tokenize(example);

tokens.forEach(convertToken);

const doc = Markdoc.parse(tokens);
const output = Markdoc.transform(doc, {});

const html = Markdoc.renderers.html(output);
console.log(html);
```

This will work, but you are probably better off implementing emoji support as an AST transform that looks for the `:x:` syntax in a Markdoc text node.
